### ISSUE TITLE
"watching 0 files" and watch detection is very slow

### ISSUE THREAD

**Issue by halt-hammerzeit** [remy/nodemon#582]
Even when it's `watching *.*`.
What's the matter with this thing?

**Comment by halt-hammerzeit** [comment]
When I delete my nodemon.json or make it an empty JSON object, it doesn't output "watching 0 files".

**Comment by halt-hammerzeit** [comment]
If my nodemon.json is like this
```
{
    verbose: true
}
```

Then it outputs "watching 0 files"

**Comment by halt-hammerzeit** [comment]
```
nodemon server/main.js --verbose --watch *.*
```

outputs "watching 0 files"

`ls` outputs a lot of files

**Comment by halt-hammerzeit** [comment]
Still it says "files triggering change check: server\main.js"
So it's lying.

**Comment by halt-hammerzeit** [comment]
And also the change detection is very slow.
It takes it about a second or two to notice the change.

**Comment by halt-hammerzeit** [comment]
And it says `[nodemon] files triggering change check: webpack.config.js` even though I didn't tell it to watch that file: I told it to watch `server/**/*.js`

**Comment by halt-hammerzeit** [comment]
Oh, it seems that after that it says `[nodemon] changes after filters (before/after): 1/0` so there is no match.
So watching patterns are not broken.
But the watching itself is very slow.

**Comment by halt-hammerzeit** [comment]
And it still watches all the directories though I didn't tell it to do so.

**Comment by halt-hammerzeit** [comment]
And why on earth does it watch the whole giant `config.dirs` folder and only when the watch triggers it checks the bloody regexp?

**Comment by halt-hammerzeit** [comment]
```
   70        var cwd = process.cwd();
   71:       if (config.dirs.length === 0) {
   72:         config.dirs.unshift(cwd);
   73        }
```

dafuq?

**Comment by halt-hammerzeit** [comment]
So this is a hidden secret "undocumented feature" and everybody suffers from that?

**Comment by halt-hammerzeit** [comment]
So, the solution:

```
config.load = function (settings, ready) {
  reset();
  var config = this;
  load(settings, config.options, config, function (options) {
    config.options = options;

    if (options.dirs instanceof Array) {
      config.dirs = options.dirs;
    }
```

And that's it.
It's that simple.

Why?

**Comment by halt-hammerzeit** [comment]
So, will you make the fix or do I have to fork the repo to is it with my projects?

**Comment by halt-hammerzeit** [comment]
I'll go make a pull request

**Comment by halt-hammerzeit** [comment]
Even with the above fix the watcher still detects changes after a second or two.

**Comment by halt-hammerzeit** [comment]
By the way, it says "watching 0 files" because there's no -L on windows

```
G:\work\react-styling [master +0 ~1 -0]> find -L "g:\work\react-styling"
"C:/Users/Nikolay/AppData/Local/GitHub/PortableGit_c2ba306e536fdf878271f7fe636a147ff37326ad/bin/find.exe": invalid predi
cate `-L'
```

You can use `-follow` instead for it to work on all platforms.

```
The -follow option has a similar effect to -L, though it takes effect at the point where it appears (that is, if -L is not used but -follow is, any symbolic links appearing after -follow on the command line will be dereferenced, and those before it will not).
```

But even then it won't work:
```
G:\work\react-styling [master +0 ~1 -0]> find -follow "server"  (  -ipath "server/.git" -prune -or  -ipath "server/node_
modules/**/*" -prune -or  -ipath "server/build/**/*" -prune ) -or  ( -type f -print ) | wc
-ipath : The term '-ipath' is not recognized as the name of a cmdlet, function, script file, or operable program. Check
 the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:27
+ find -follow "server"  (  -ipath "server/.git" -prune -or  -ipath "server/node_m ...
+                           ~~~~~~
    + CategoryInfo          : ObjectNotFound: (-ipath:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

**Comment by halt-hammerzeit** [comment]
My previous pull request got messed up so here's the correct one
https://github.com/remy/nodemon/pull/584

**Comment by halt-hammerzeit** [comment]
By the way, I can see now why you might be watching the whole project folder.
This way you can detect more new files and new subfolders.
I don't know how much the performance penalty might be.
I guess it may be negligible.
But in some cases, assume something like a web downloader, it might get high (when the program constantly writes to disk).
If you think this kind of blind optimization is not worthy you can drop the pull request, I'm not an expert here.
But you better reflect this undocumented feature in the Readme because everyone thinks you just watch the regexped folders.

**Comment by remy** [comment]
Please try the build here #625 and feedback against the issue #625 (I'm trying to refactor the whole watch logic out of nodemon entirely!).

**Comment by halt-hammerzeit** [comment]
@remy i've installed 1.5.0 and i'm not having the issue. And while I was having an older version I didn't face the issue too. I'm not sure what changed, but currently I've no complaints on how nodemon watches files.

**Comment by remy** [comment]
@halt-hammerzeit note that 1.5 is an alpha dev build, so you'll need to (or would want to) upgrade again, but hopefully the update-notifier should help you along anyway. The watch methods have been completely gutted which should help with these issues.

**Comment by remy** [comment]
#625 closes and released in `nodemon@1.4.2`
